clang/runtime/dpct-rt/include/dpct/blas_utils.hpp

```diff
@@ -2467,8 +2467,777 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
+  case library_data_t::real_half:
+    return dnnl::memory::data_type::f16;
+  case library_data_t::real_bfloat16:
+    return dnnl::memory::data_type::bf16;
+  case library_data_t::real_float:
+    return dnnl::memory::data_type::f32;
+  case library_data_t::real_double:
+    return dnnl::memory::data_type::f64;
+  case library_data_t::real_int32:
+    return dnnl::memory::data_type::s32;
+  case library_data_t::real_int8:
+    return dnnl::memory::data_type::s8;
+  case library_data_t::real_uint8:
+    return dnnl::memory::data_type::u8;
+  default:
+    throw std::runtime_error(
+        "the input library_data_t type has no corresponding type in dnnl");
+  }
+}
+
+template <class T, class Talpha>
+sycl::event scale_d_with_vector_alpha_impl(queue_ptr q_ptr, int rows, int cols,
+                                           T *d, const Talpha *alpha) {
+  return q_ptr->submit([&](sycl::handler &cgh) {
+#ifdef DPCT_USM_LEVEL_NONE
+    access_wrapper<T *> d_acc(d, cgh);
+    access_wrapper<const Talpha *> alpha_acc(alpha, cgh);
+#endif
+    cgh.parallel_for<
+        dpct_kernel_name<class scale_with_vector_alpha, T, Talpha>>(
+        sycl::range<2>(rows, cols), [=](sycl::id<2> index) {
+#ifdef DPCT_USM_LEVEL_NONE
+          auto d_data = d_acc.get_raw_pointer();
+          auto alpha_data = alpha_acc.get_raw_pointer();
+#else
+            auto d_data = d;
+            auto alpha_data = alpha;
+#endif
+          size_t row_idx = index.get(0);
+          size_t col_idx = index.get(1);
+          size_t idx = rows * col_idx + row_idx;
+          d_data[idx] = d_data[idx] * alpha_data[row_idx];
+        });
+  });
+}
+
+// d is col major without padding
+inline sycl::event scale_d_with_vector_alpha(queue_ptr q_ptr, int rows,
+                                             int cols, void *d,
+                                             library_data_t d_type,
+                                             const void *alpha,
+                                             library_data_t alpha_type) {
+  std::uint64_t key = dpct::detail::get_type_combination_id(d_type, alpha_type);
+  sycl::event e;
+  switch (key) {
+  case dpct::detail::get_type_combination_id(library_data_t::real_int8,
+                                             library_data_t::real_float): {
+    e = scale_d_with_vector_alpha_impl<std::int8_t, float>(
+        q_ptr, rows, cols, (std::int8_t *)d, (const float *)alpha);
+    break;
+  }
+  case dpct::detail::get_type_combination_id(library_data_t::real_int32,
+                                             library_data_t::real_float): {
+    e = scale_d_with_vector_alpha_impl<int, float>(q_ptr, rows, cols, (int *)d,
+                                                   (const float *)alpha);
+    break;
+  }
+  default:
+    throw std::runtime_error("the combination of data type is unsupported");
+  }
+  return e;
+}
+
+inline std::tuple<size_t, size_t>
+get_from_to_linear_idx(size_t rows, size_t cols, size_t a_ld, order_t a_order,
```

Comment:
// get a linear idx map for a 2D point (row_idx, col_idx) between src_order and dst_order
get_linear_idx_map(size_t rows, size_t cols, size_t src_ld, order_t src_order,
                       size_t dst_ld, order_t dst_order, size_t row_idx,
                       size_t col_idx)

Reply:
done